### PR TITLE
[iOS] Fix for CollectionView leaving behind old items

### LINKED ISSUE #5831: [iOS] Layout and Styling Issues when reloading page
### Description

Item from `CollectionView` is left over or out of place. Title styling is lost.

https://www.screencast.com/t/zB074v71xe

Android doesn't exhibit these problems, but the scrolling is very choppy and slow on the emulator, even after the images are all loaded.

https://www.screencast.com/t/wm7OZVLWW

### Details

- Version with issue: 4.0-pre8

### Reproduction Link

https://github.com/davidortinau/Gastropods

### LINKED ISSUE #6507: [Bug] Navigating away from CollectionView and coming back leaves weird old items
### Description

On iOS, when using a CollectionView in Shell, navigating away from it with `Shell.GoToAsync()` and going back makes the old items appear stacked together, while the new items are displayed correctly.

### Steps to Reproduce

1. Clone the Xaminals sample (https://github.com/xamarin/xamarin-forms-samples/blob/master/UserInterface/Xaminals)
2. Debug the iOS project
3. Check that the CollectionView of cats appears correctly
4. Open the flyout menu
5. Tap "About"
6. Repoen the flyout menu
7. Tap "Domestic"
8. Notice that the CollectionView now displays 2 lists of cats, one correct and one stack together on one line

### Expected Behavior

The CollectionView should appear correctly on both cases

### Actual Behavior

The CollectionView appears incorrectly

### Basic Information

- Version with issue: 4.0.0.425677
- IDE: Visual Studio for Mac
- Platform Target Frameworks:
  - iOS:  12.2

### Screenshots

Simulator Screen Shot - iPhone Xʀ - 2019-06-12 at 20 30 54

### Reproduction Link

https://github.com/xamarin/xamarin-forms-samples/blob/master/UserInterface/Xaminals/

### PR DESCRIPTION
**Pull request by alanag13**
### Description of Change ###
Fixes an issue where old collection view items stayed rendered across subsequent views of a page, causing a floating/piling effect of those items. This is done by explicitly clearing the data source and refreshing the `UICollectionView` before disposing it.

### Issues Resolved ### 

- fixes #5831 
- fixes #6507

### Platforms Affected ### 
- iOS

### Behavioral/Visual Changes ###

Users should no longer experience visual artifacting of prior CollectionView items floating above their exsting ones.

### Before/After Screenshots ### 

Will add, see my comment

### Testing Procedure ###

Will add, see my comment

### PR Checklist ###

- [ ] Has automated tests 
- [X] Rebased on top of the target branch at time of PR
- [X] Changes adhere to coding standard

### CODE REVIEW

**Comment by alanag13** [comment]
I've only ever seen this issue in a Shell app and haven't done a Shell PR before. Where/how do I add visual issue repros for shell to the project?

**Comment by jfversluis** [comment]
@alanag13 have a look at [Issue4684](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue4684.xaml), [4992](https://github.com/xamarin/Xamarin.Forms/tree/master/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared) or [5131](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5131.cs) or [5132](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5132.cs)

Does that help? If not, please let me know. Also there is a conflict that needs to be fixed. Thanks!

File: Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewController.cs [hunk 1]
```diff
@@ -68,6 +68,13 @@ protected override void Dispose(bool disposing)
 				if (disposing)
 				{
 					_itemsSource?.Dispose();
+					_itemsSource = null;
+
+					//clear data
+					CollectionView.DataSource = null;
+					CollectionView.ReloadData();
```
Comment (paymicro) [inline Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewController.cs]: Perhaps after it is worth adding `CollectionView.CollectionViewLayout.InvalidateLayout();` Like as in `UpdateItemsSourceAndReload()` function.

**Comment by alanag13** [comment]
@samhouts @jfversluis @paymicro This is tagged as `retarget-required` but also in review for master. Should I keep as is or retarget to 4.1?

**Comment by samhouts** [comment]
@alanag13 master is perfect. thanks!!

**Comment by jfversluis** [comment]
@alanag13 Did you see our other comments? Can you work with that? If you need anything please let us know

**Comment by alanag13** [comment]
@jfversluis @paymicro I should be able to add the tests and feedback for this one tonight.

**Comment by jfversluis** [comment]
Hey @alanag13 where are we on the tests, etc.? :)

**Comment by jfversluis** [comment]
Failing test is actually a succeeded one but somehow the connection to App Center was interrupted

**Comment by alanag13** [comment]
@jfversluis sorry for the delay here, was at the conference last week. I'll try to button this one soon, sorry to keep you waiting

**Comment by jfversluis** [comment]
Hey @alanag13 no problem at all! Don’t mistake my comments for being impatient or trying to push you. We understand that you’re doing this in your spare time and appreciate that very much!

Especially the “changes requested” is mostly so that this PR doesn’t keep showing up in an internal alert system that I still need to review this while I am actually waiting for an action of the author. So it was mostly that I was done with the alerts 😬 

Whenever you’re ready let me know and I’ll be happy to have another look. I hope your conference was great!

**Comment by alanag13** [comment]
@jfversluis addressed feedback and added tests.

File: Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewController.cs (hunk 1 above)
Comment (alanag13) [reply Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewController.cs]: Thanks, added.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5831.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#endif
+
+namespace Xamarin.Forms.Controls.Issues
+{
+#if UITEST
+	[Category(UITestCategories.ManualReview)]
+#endif
+	[Preserve(AllMembers = true)]
+	[Issue(IssueTracker.Github, 5831, "Navigating away from CollectionView and coming back leaves weird old items", PlatformAffected.iOS)]
+	public class Issue5831 : TestShell
```
Comment (jfversluis) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5831.cs]: Don’t we also need an entry in the projitems for this to be included?

Comment (alanag13) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5831.cs]: 🤦‍♂️ not sure how that got excluded. I'll get hat fixed

Comment (alanag13) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue5831.cs]: @jfversluis sorry for the delay, I've finally added this now!

**Comment by alanag13** [comment]
@jfversluis these were passing before, are the failing UITests related? Let me know if there's anything for me to do, I'll try to turn things arounds a little faster this time...

**Comment by jfversluis** [comment]
The tests seem a little unpredictable at the moment unfortunately. I have checked them by hand and the failures seem unrelated. LGTM! :)

**Comment by jfversluis** [comment]
@samhouts was there any particular reason the do-not-merge tag was on here? Or just because there was some feedback to be processed that is now implemented?

**Comment by samhouts** [comment]
@jfversluis Nope, that was it!

**Comment by hartez** [comment]
I'm not seeing this fix work when I test it; if I build check this out and run it, I'm still seeing the leftover text on the first ContentPage. I've tried this against iOS 11 and 12, same results.

**Comment by hartez** [comment]
> I've only ever seen this issue in a Shell app

This makes me wonder if the issue is a Shell problem, rather than a CollectionView problem. The mechanism behind this fix doesn't make a lot of sense - if the CollectionView and its renderers/controllers are all property cleaned up, then we shouldn't have to "clear" the data to stop them from showing up. I suspect that some part of the CollectionView is still hanging around, and this fix is merely hiding it.

**Comment by alanag13** [comment]
@hartez that makes sense, I'll take a closer look. I'm a little surprised that youre seeing this not fix anything  (I built a private version of the XF nuget that works for my app) but if it's not the "right" fix anyway then more exploring should be done.

**Comment by jfversluis** [comment]
We're seeing this issue for CarouselView as well. I can confirm that this code does not fix the issue. However, it's also not something that is specific to Shell as it also happens in the Gallery app

**Comment by alanag13** [comment]
That's insightful, thanks. I'll try to take another peek at this soon. I wonder if something else changed that caused this to no longer work once it was rebased because I do have an older build with this code where it *does* solve it (even if it's not the right *right* solve as stated above).

**Comment by jfversluis** [comment]
Here is a screen recording: https://1drv.ms/u/s!Anrk6U9nRknzgvxoOdpDfJqc8sHm2g

This code should be available soon for you to try yourself as well. The funny thing is that it does seem to be of influence which item is selected when you leave the page and come back. If you look very closely on the "Position View" one you can see that the actual carousel behind it is moving (watch the corners). It's more clear from the "PeakAreaInsets" view. Then when you go back and forth between page you can see that the stuck cell (at least the top one, maybe there are more?) changes value.

I've been doing some diving in this, but no clear resolution yet.

**Comment by hartez** [comment]
PR #8189 is the correct fix for this, I think. Waiting for the automated test results to confirm. 
Also, I pulled in your UI test commit - it's by far the easiest repro of the problem I've seen. 

If 8189 doesn't end up solving the issue, I'll open this one back up and we can go from there.

**Pull request closed without merging**